packages/react/scroll-area/src/ScrollArea.tsx

```diff
@@ -186,2234 +80,886 @@ type ScrollAreaPrimitive = Polymorphic.ForwardRefComponent<
-          scrollbarVisibility={scrollbarVisibility}
-          scrollbarVisibilityRestTimeout={scrollbarVisibilityRestTimeout}
-          trackClickBehavior={trackClickBehavior}
-        >
-          <ScrollAreaStateContext.Provider value={reducerState}>
-            <Primitive
-              {...domProps}
-              ref={ref}
-              style={{
-                ...domProps.style,
-                ...style,
-              }}
-              onPointerEnter={composeEventHandlers(props.onPointerEnter, onPointerEnter)}
-              onPointerLeave={composeEventHandlers(props.onPointerLeave, onPointerLeave)}
-            />
-          </ScrollAreaStateContext.Provider>
-        </ScrollAreaProvider>
-      </ScrollAreaRefsProvider>
-    </DispatchProvider>
+    <Presence present={forceMount || state !== 'hidden'}>
+      <ScrollAreaScrollbarVisible
+        {...scrollbarProps}
+        ref={forwardedRef}
+        onPointerEnter={composeEventHandlers(props.onPointerEnter, () => send('POINTER_ENTER'))}
+        onPointerLeave={composeEventHandlers(props.onPointerLeave, () => send('POINTER_LEAVE'))}
+      />
+    </Presence>
   );
-}) as ScrollAreaImplPrimitive;
+}) as ScrollAreaScrollbarOptionalPrimitive;
+
+const ScrollAreaScrollbarAuto = React.forwardRef((props, forwardedRef) => {
+  const context = useScrollAreaContext(SCROLLBAR_NAME);
+  const { forceMount, ...scrollbarProps } = props;
+  const [visible, setVisible] = React.useState(false);
+  const isHorizontal = props.orientation === 'horizontal';
+
+  useResizeObserver(context.viewport, () => {
+    if (context.viewport) {
+      const isOverflowX = context.viewport.offsetWidth < context.viewport.scrollWidth;
+      const isOverflowY = context.viewport.offsetHeight < context.viewport.scrollHeight;
+      setVisible(isHorizontal ? isOverflowX : isOverflowY);
+    }
+  });
 
-/* -------------------------------------------------------------------------------------------------
- * ScrollAreaViewport
- * -----------------------------------------------------------------------------------------------*/
+  return (
+    <Presence present={forceMount || visible}>
+      <ScrollAreaScrollbarVisible {...scrollbarProps} ref={forwardedRef} />
+    </Presence>
+  );
+}) as ScrollAreaScrollbarOptionalPrimitive;
 
-const VIEWPORT_NAME = 'ScrollAreaViewport';
+/* -----------------------------------------------------------------------------------------------*/
 
-type ScrollAreaViewportImplOwnProps = Polymorphic.OwnProps<typeof Primitive>;
-type ScrollAreaViewportImplPrimitive = Polymorphic.ForwardRefComponent<
-  Polymorphic.IntrinsicElement<typeof Primitive>,
-  ScrollAreaViewportImplOwnProps
+type ScrollAreaScrollbarVisibleOwnProps = Polymorphic.Merge<
+  Omit<
+    | Polymorphic.OwnProps<typeof ScrollAreaScrollbarX>
+    | Polymorphic.OwnProps<typeof ScrollAreaScrollbarY>,
+    keyof ScrollAreaScrollbarAxisPrivateProps
+  >,
+  { orientation?: 'horizontal' | 'vertical' }
```

Comment:
This was actually just me being lazy because both components have the same props so it doesn't really matter. You're right though that I would need to make `orientation` a required prop if we wanted to do a more accurate union type here 😬 

I did wonder if `orientation` should be required though... What do you think?

Reply:
Seems fine that `orientation` here would be required, I guess on the flipside, it's nice to be able to throw in a scrollbar when you just want vertical and have it just work… I'm torn.

Reply:
> it's nice to be able to throw in a scrollbar when you just want vertical and have it just work

Haha exactly... this was my thinking but also torn. Shall we leave as is for now?

Reply:
Yeah let's leave it for now.